### PR TITLE
Ubisys D1 - Add onOffTransitionTime and startUpCurrentLevel support

### PR DESCRIPTION
**Pull request by sjorge**
Depends on https://github.com/Koenkk/zigbee-herdsman/pull/283

- [x] testing
- [x] exposes

I don't have the device to test, I saw https://github.com/Koenkk/zigbee2mqtt/discussions/5555 and it looks like it's just standard ZCL so I added it.

@Koenkk not sure how I feel about the cameCase inside level_config, I kept it to simple match what we define in herdsman, if we ever allow more values we can easily add it. Looks similar to than what ballast_config did.

### CODE REVIEW

**Comment by sjorge** [comment]
@Koenkk any idea if this would say work with a regular bulb? Then maybe I can test it myself.

**Comment by sjorge** [comment]
# Testing
Looks like the older FLAOLT panels do support it! I can test!

## read the attribute
```
Zigbee2MQTT:debug 2021-01-08 18:54:29: Publishing get 'get' 'level_config' to 'office/light/main/panel1'
Zigbee2MQTT:debug 2021-01-08 18:54:29: Received Zigbee message from 'office/light/main/panel1', type 'readResponse', cluster 'genLevelCtrl', data '{"onOffTransitionTime":5,"startUpCurrentLevel":254}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2021-01-08 18:54:32: MQTT publish: topic 'zigbee2mqtt/office/light/main/panel1', payload '{"brightness":62,"color":{"x":0.5018,"y":0.4153},"color_temp":454,"level_config":{"onOffTransitionTime":5,"startUpCurrentLevel":254},"linkquality":42,"state":"OFF","update":{"state":"idle"}}'
```
## writing the attribute
TBD

**Comment by sjorge** [comment]
@Koenkk I reworked it to grab the other 2 attributes too but ran into an issue

```
Zigbee2MQTT:error 2021-01-08 19:54:03: Publish 'get' 'level_config' to 'office/light/main/panel1' failed: 'Error: Read 0x14b457fffe2fee7a/1 genLevelCtrl(["onOffTransitionTime","onTransitionTime","offTransitionTime","startUpCurrentLevel"], {"timeout":10000,"disableResponse":false,"disableReco
very":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null}) failed (Status 'UNSUPPORTED_ATTRIBUTE')'                                                                                                     Zigbee2MQTT:debug 2021-01-08 19:54:03: Error: Read 0x14b457fffe2fee7a/1 genLevelCtrl(["onOffTransitionTime","onTransitionTime","offTransitionTime","startUpCurrentLevel"], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint
":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null}) failed (Status 'UNSUPPORTED_ATTRIBUTE')
    at Endpoint.checkStatus (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/dist/controller/model/endpoint.js:185:23)
    at Endpoint.<anonymous> (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/dist/controller/model/endpoint.js:240:26)
    at Generator.next (<anonymous>)
    at fulfilled (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/dist/controller/model/endpoint.js:24:58)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

Looks like when not all four are support it throws this... how can try to read all four but be OK if say only 2 are supported?

**Comment by sjorge** [comment]
# Testing

## Reading FLOALT panel

```
Zigbee2MQTT:debug 2021-01-08 20:09:13: Received MQTT message on 'zigbee2mqtt/office/light/main/panel1/get/level_config' with data ''
Zigbee2MQTT:debug 2021-01-08 20:09:13: Publishing get 'get' 'level_config' to 'office/light/main/panel1'
Zigbee2MQTT:debug 2021-01-08 20:09:13: Received Zigbee message from 'office/light/main/panel1', type 'readResponse', cluster 'genLevelCtrl', data '{"onOffTransitionTime":5}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-08 20:09:13: Received Zigbee message from 'office/light/main/panel1', type 'readResponse', cluster 'genLevelCtrl', data '{}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-08 20:09:13: Received Zigbee message from 'office/light/main/panel1', type 'readResponse', cluster 'genLevelCtrl', data '{}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-08 20:09:13: Received Zigbee message from 'office/light/main/panel1', type 'readResponse', cluster 'genLevelCtrl', data '{"startUpCurrentLevel":254}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2021-01-08 20:09:16: MQTT publish: topic 'zigbee2mqtt/office/light/main/panel1', payload '{"brightness":254,"color":{"x":0.5018,"y":0.4153},"color_temp":454,"level_config":{"onOffTransitionTime":5,"startUpCurrentLevel":254},"linkquality":42,"onOffTransitionTime":0,"startUpCurrentLevel":"max","state":"OFF","update":{"state":"idle"}}'
```

**Comment by Koenkk** [comment]
> Looks like when not all four are support it throws this... how can try to read all four but be OK if say only 2 are supported?

This is not possible indeed. I suggest to split up the read (also reading 4 attributes at the same time fails for some Hue devices, 2 - 3 is the max)

**Comment by sjorge** [comment]
So far it seems most Trådfri seems to support `onOffTransitionTime` and `startUpCurrentLevel`.
Innr and Tint seem to support only `startUpCurrentLevel`. I don't have my other test bulbs paired at the moment.

So far none of them seem to support `onTransitionTime` and `offTransitionTime`... but all 4 are marked optional in ZCL so that is to be expected I guess. But hopefully it will help https://github.com/Koenkk/zigbee2mqtt/discussions/5555 and others. I currently have onoffTransitionTime set to 0 on my office FLOATS and they now match the instance on/off form my mood lighting (Innr) shame I can't set the Innr's to 5... but oh well. (Oddly enough, Tradfri is usually the most correct in all these things, no lying in colorCaps, support most optional attributes, ...)

image

File: converters/fromZigbee.js [hunk 1]
```diff
@@ -430,6 +430,53 @@ const converters = {
...
     },
+    level_config: {
+        cluster: 'genLevelCtrl',
+        type: ['attributeReport', 'readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            const result = {'level_config': {}};
+
+            // onOffTransitionTime - range 0x0000 to 0xffff - optional
+            if (msg.data.hasOwnProperty('onOffTransitionTime') && (msg.data['onOffTransitionTime'] !== undefined)) {
+                result.level_config.onOffTransitionTime = msg.data['onOffTransitionTime'];
```
Comment (Koenkk) [inline converters/fromZigbee.js:441]: Maybe `level` should be renamed to `brightness` here, I guess users aren't familiar with the `level` term.

File: lib/exposes.js [hunk 2]
```diff
@@ -177,6 +177,32 @@ class Light extends Base {
         return this;
     }
 
+    withLevelConfig() {
```
Comment (Koenkk) [inline lib/exposes.js:180]: Maybe we should only add this for bulbs we know it works with. I can already imagine the amount of issues being created because these won't work for China bulb X

Comment (sjorge) [reply lib/exposes.js:180]: What about not exposing it for bulbs at all, but exposing it for devices like the Ubisys D1?
I imagine you want to use this on a D1, but probably not on bulbs? (But it's nice to have the converters hooked up in case you do)

Comment (Koenkk) [reply lib/exposes.js:180]: I agree, converters are indeed fine and nothing against in having these by default. Exposes only for D1 sounds good

File: converters/fromZigbee.js (hunk 1 above)
Comment (sjorge) [reply converters/fromZigbee.js:441]: I've now removed it from the expose for bulbs and manually added it for the Ubisys D1.

I kind of want to leave it at level_config, as the tech manual the user was using also refers to it as the level cluster... and tbh, if you want to use this, you probably know what you're doing and we should dumb it down?

**Comment by sjorge** [comment]
Quickly added my single Hue and Osram bulb, osram supports none of this. Hue only startUpCurrentLevel.

Ikea surprisingly on top with 2/4, Innr+Tint+Hue 1/4, Osram 0/4 ... for the best we don't add this for bulbs.

File: converters/fromZigbee.js (hunk 1 above)
Comment (Koenkk) [reply converters/fromZigbee.js:441]: Agree

File: converters/fromZigbee.js [hunk 3]
```diff
@@ -430,6 +430,53 @@ const converters = {
...
+            }
+
+            // onTransitionTime - range 0x0000 to 0xffff - optional
+            //                    0xffff = use onOffTransitionTime
+            if (msg.data.hasOwnProperty('onTransitionTime') && (msg.data['onTransitionTime'] !== undefined)) {
+                result.level_config.onTransitionTime = msg.data['onTransitionTime'];
+                if (result.level_config.onTransitionTime == 65535) {
```
Comment (Koenkk) [inline converters/fromZigbee.js:448]: Can you make all published properties snake  case (e.g. `on_transition_time`)?

File: converters/toZigbee.js [hunk 4]
```diff
@@ -239,6 +239,93 @@ const converters = {
...
+    level_config: {
+        key: ['level_config'],
+        convertSet: async (entity, key, value, meta) => {
+            const state = {};
+
+            // parse payload to grab the keys
+            try {
+                value = JSON.parse(value);
+            } catch (e) {
+                e;
```
Comment (Koenkk) [inline converters/toZigbee.js:251]: I think we should throw an error in this case.

Comment (sjorge) [reply converters/toZigbee.js:251]: So just remove the try-catch? Or leave as is but log an error via meta.logger ?

File: converters/fromZigbee.js (hunk 3 above)
Comment (sjorge) [reply converters/fromZigbee.js:448]: Sure, i think I saw a function to translate snake to camel somewhere in zhc recently, let me poke around.

File: converters/toZigbee.js (hunk 4 above)
Comment (Koenkk) [reply converters/toZigbee.js:251]: throw a friendly error here, e.g. `throw new Error("Payload is not valid JSON")`

**Comment by sjorge** [comment]
Found an issue after retesting with the snake case conversion, not sure what it is yet but after a set... I get no response.

**Comment by sjorge** [comment]
OK figure it out, will fix before heading to bed.

**Comment by Koenkk** [comment]
Thanks!
